### PR TITLE
End-to-end test coverage for CLI commands output

### PR DESCRIPTION
**Pull request by hidmic**
This pull request aims to add full end-to-end test coverage to ~all~ several `ros2` CLI commands:

|                      | topic | service | action | ~param~ | msg | srv | interface | node | ~run~ | ~component~ | ~lifecycle~ | pkg | ~daemon~ | ~launch~ |
|----------------------|:-----:|:-------:|:------:|:-------:|:---:|:---:|:---------:|:----:|:---:|:---------:|:---------:|:---:|:----:|:----:|
| Correct output       |   ✓   |    ✓    |    ✓   |   ✗   |  ✓  |  ✓  |     ✓     |   ✓  |  ✗  |     ✗     |     ✗     |  ✓  |  ✗  |  ✗  |
| Graceful termination |   ✓   |    ✓    |    ✓   |   ✗   |  ✓  |  ✓  |     ✓     |   ✓  |  ✗  |     ✗     |     ✗     |  ✓  |  ✗  |  ✗  |
| Error handling       |   ✓   |    ✓    |    ✓   |   ✗   |  ✓  |  ✓  |     ✓     |   ✓  |  ✗  |     ✗     |     ✗     |  ✓  |  ✗  |  ✗  |

~Note: this table will be updated as progress is made.~ Test coverage for commands that have been struck through i.e. `param`, `component`, `lifecycle`, etc., will not be provided here but in follow-up PRs. These commands have side effects beyond their execution lifetime that have to be checked in ways that are specific to each command. Thus, these don't lend themselves well to be integrated with the "standardized" testing device used for other commands.

Connected to https://github.com/ros2/ros2/issues/739.

### CODE REVIEW

**Comment by hidmic** [comment]
There's still the discussion as to whether this is enough for end-to-end testing, let alone unit tests which are not covered by this PR.

**Comment by hidmic** [comment]
Running CI to uncover non-Linux, platform specific issues (build up to test_ros2cli, test test_ros2cli, including opensplice):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7618/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3736/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6229/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7444/)

**Comment by hidmic** [comment]
Ok, I'll stop here for the reasons exposed in [this PR's description](https://github.com/ros2/ros2cli/pull/304#issue-299381542) and because this is getting larger than I'm usually comfortable with.

File: ros2node/ros2node/verb/list.py [hunk 1]
```diff
@@ -37,4 +37,4 @@ def main(self, *, args):
         if args.count_nodes:
             print(len(node_names))
         elif node_names:
-            print(*[n.full_name for n in node_names], sep='\n')
+            print(*sorted(n.full_name for n in node_names), sep='\n')
```
Comment (dirk-thomas) [inline ros2node/ros2node/verb/list.py]: This change as well as the ones in the `bw` module are functional changes not related to testing. Please move them into a separate PR. Otherwise when the PR is being squash merged they are "hidden" in a commit only mentioning "test coverage".

File: test_ros2cli/package.xml [hunk 2]
```diff
@@ -12,11 +12,21 @@
   <test_depend>action_tutorials</test_depend>
   <test_depend>ament_lint_auto</test_depend>
   <test_depend>ament_lint_common</test_depend>
+  <test_depend>demo_nodes_py</test_depend>
```
Comment (dirk-thomas) [inline test_ros2cli/package.xml]: This will create a repository level cycle since packages in the repository containing `demo_nodes_py` depend on packages in this repo. Imo we shouldn't do that.

File: test_ros2cli/CMakeLists.txt [hunk 3]
```diff
@@ -19,14 +19,35 @@ if(BUILD_TESTING)
...
+    if(NOT DEFINED TEST_CONFIG_MODULE)
+      set(TEST_CONFIG_MODULE ${test_name}_config)
+    endif()
+    # Setup delay for test fixture actions
+    if(NOT DEFINED TEST_SETUP_DELAY)
+      set(TEST_SETUP_DELAY 2.0)
+      if("${TEST_RMW_IMPLEMENTATION}" STREQUAL "rmw_connext_cpp")
+        # Connext startup is too slow. It needs a few extra seconds till
+        # discovery completes.
+        set(TEST_SETUP_DELAY 5.0)
```
Comment (dirk-thomas) [inline test_ros2cli/CMakeLists.txt]: That looks problematic because it:
* is fragile in case Connext might just take a second longer
* is slowing down the tests in case Connext happens to be faster than the threshold
* needs to be done for potentially other RMW impl

Instead we should find a way to write the test to work without this delay.

File: test_ros2cli/CMakeLists.txt [hunk 4]
```diff
@@ -19,14 +19,35 @@ if(BUILD_TESTING)
...
+      endif()
+    endif()
+    # Timeout for test itself
+    if(NOT DEFINED TEST_TIMEOUT)
+      set(TEST_TIMEOUT 0)
```
Comment (dirk-thomas) [inline test_ros2cli/CMakeLists.txt]: A timeout of `zero` doesn't sound like a good idea. Please always use an upper boundary.

File: test_ros2cli/package.xml (hunk 2 above)
Comment (hidmic) [reply test_ros2cli/package.xml]: Ok, that is true for `ros2run`. It's another instance of an issue we have elsewhere e.g. `launch_ros`, and that I was consciously eluding to avoid scope creep. IMO we shouldn't be using demo nodes for system or integration testing, neither duplicate `ros2/system_tests` talkers, listeners, servers, etc., and others. We need a package with test nodes that `ros2/ros2cli`, `ros2/launch*`, `ros2/system_tests`, can all depend on. Probably in its own repo, like we did for `test_interface_files`.

I can put this on hold, do that and come back, or fix it afterwards. What would you rather do?

File: ros2node/ros2node/verb/list.py (hunk 1 above)
Comment (hidmic) [reply ros2node/ros2node/verb/list.py]: Oops, yes, you're right. I'll split it up.

File: test_ros2cli/CMakeLists.txt (hunk 3 above)
Comment (hidmic) [reply test_ros2cli/CMakeLists.txt]: It is problematic, I know. Truth to be told, all these tests have a delay to deal with CPU load fluctuations, uneven discovery times, etc., which makes them all flakey. I had to remove publishing/echoing rate tests because it's completely inconsistent, time wise.

Now, to fix this. Many tests don't need this delay at all e.g. `ros2 msg list`, so we can cut down time there. For those that do need it, I see two kinds: those that are bound to communication latency and those that are bound to discovery time. 

For the first kind, we can wait for specific output or events e.g. for `ros2 topic bw` wait till we have _any_ (non filtered) output to start testing it. 

For the second kind e.g. `ros2 topic list`, well, it gets harder. Assuming the daemon is there, we may repeatedly execute the command, though that isn't really less fragile than a delay (and I can't figure how to pull that off using `ros_testing` as it is today). We may also wait for the ROS graph to show some nodes, but since the CLI command, the CLI daemon and launch test are three separate processes, it may not work as intended.

Alternatively, I think that we need to make our CLI more predictable. Just like `ros2 service call` awaits for the server to be available, `ros2 service list` could have an optional ROS graph condition to wait for e.g. `ros2 service list --when "'my_server' in nodes"`. The same goes for all discovery time bound commands. What do you think?

**Comment by hidmic** [comment]
> The size of the tests makes me wonder if we really can't write them more compact?

What do you mean by size? These tests leverage the commonalities in the CLI commands to have single `test_cli.py.in` instance per command, parameterized N times. Thus the logic in `test_cli.py.in`. We could find other ways to reuse and instead have N tests. That'll yield shorter tests but the sheer size of this PR will increase due to boilerplate code duplication. Is that what you had in mind @dirk-thomas ?

> What is the reason that all the tests are bundled in the separate package test_ros2cli?

There's currently no way to setup launch tests in pure Python packages (see https://github.com/ros2/launch/issues/237).

File: test_ros2cli/CMakeLists.txt (hunk 4 above)
Comment (hidmic) [reply test_ros2cli/CMakeLists.txt]: Completely agree. Changed in https://github.com/ros2/ros2cli/pull/304/commits/08d5088fda6846af14aff38915e227336e1c6819.

File: test_ros2cli/package.xml (hunk 2 above)
Comment (dirk-thomas) [reply test_ros2cli/package.xml]: Imo a package/repo like `test_nodes` sounds like a good idea.

File: test_ros2cli/CMakeLists.txt (hunk 3 above)
Comment (dirk-thomas) [reply test_ros2cli/CMakeLists.txt]: > Assuming the daemon is there, we may repeatedly execute the command, though that isn't really less fragile than a delay.

Why would that not be less fragile? It will pass the test as soon as the command returns the desired output and only runs to a maximum timeout if the command doesn't do so.

> Just like `ros2 service call` awaits for the server to be available, `ros2 service list` could have an optional ROS graph condition to wait for e.g. `ros2 service list --when "'my_server' in nodes"`.

This is a good example how *not* to add conditions imo. If you want to wait for X (services of a node being available) you can't use a condition Y (check if a node has been discovered). Simply because there is a still a race since the node might become available first shortly before the services are actually advertised.

File: ros2node/ros2node/verb/list.py (hunk 1 above)
Comment (hidmic) [reply ros2node/ros2node/verb/list.py]: See #305 and #306.

File: test_ros2cli/test/test_cli.py.in [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        sigterm_timeout=LaunchConfiguration(
+            'sigterm_timeout', default=60
+        )
+    )
+    actions = [command_under_test]
+    actions.append(
+        OpaqueFunction(function=lambda context: ready_fn())
+    )
+    launch_description = LaunchDescription([
+        # Always restart daemon to isolate tests.
```
Comment (dirk-thomas) [inline test_ros2cli/test/test_cli.py.in]: Maybe instead of doing this it would be simpler to add an option `--no-daemon` to all the commands to not start the daemon on demand?

File: test_ros2cli/CMakeLists.txt [hunk 6]
```diff
@@ -52,17 +73,40 @@ if(BUILD_TESTING)
...
+      CONFIG_MODULE ros2msg_test_configuration
+      TIMEOUT 240)
+    add_custom_cli_test(
+      test_ros2srv
+      CONFIG_MODULE ros2srv_test_configuration
+      TIMEOUT 240)
+    add_custom_cli_test(
+      test_ros2interface
+      CONFIG_MODULE ros2interface_test_configuration
+      TIMEOUT 240)
```
Comment (dirk-thomas) [inline test_ros2cli/CMakeLists.txt]: The sum of all timeouts is 31 min. Is that really necessary?

File: test_ros2cli/CMakeLists.txt (hunk 3 above)
Comment (hidmic) [reply test_ros2cli/CMakeLists.txt]: > Why would that not be less fragile? It will pass the test as soon as the command returns the desired output and only runs to a maximum timeout if the command doesn't do so.

I meant neither option makes it not fragile. You may not delay enough or you may not wait enough. Sure, we can increase the timeout a lot to increase the likelihood of discovery to find all peers for free, but that doesn't make it not fragile. If CI nodes are running with enough load, tests may still flake.

Anyways, even if we go down that route, we lack support in launch testing to do it so we have to go back to it first.

> Simply because there is a still a race since the node might become available first shortly before the services are actually advertised.

Yes, it doesn't make it not fragile either. I should've said that. I'm not well versed enough on DDS discovery mechanisms, is it atomic in any aspect e.g. some information chunks that are guaranteed to arrive in single block?

Comment (dirk-thomas) [reply test_ros2cli/CMakeLists.txt]: > Sure, we can increase the timeout a lot to increase the likelihood of discovery to find all peers for free, but that doesn't make it not fragile. If CI nodes are running with enough load, tests may still flake.

I don't think that is the definition of flaky. The fact that we have to define an upper bound how long we are willing to wait is fine.

File: test_ros2cli/test/test_cli.py.in (hunk 5 above)
Comment (hidmic) [reply test_ros2cli/test/test_cli.py.in]: Hmm, that is a good idea, but won't it result in test instances affecting each other through daemon state?

File: test_ros2cli/CMakeLists.txt (hunk 6 above)
Comment (hidmic) [reply test_ros2cli/CMakeLists.txt]: It's insane. I was going to bring it up in tomorrow's meeting. And yes, it used to be half of it and tests would timeout in CI. Cutting down delays we may be able to reduce it, but I wouldn't expect a large reduction e.g. Connext is **very** slow.

**Comment by dirk-thomas** [comment]
> What do you mean by size?

I was wondering why are `TestCommandFinishesInAFiniteAmountOfTime` as well as `TestCommandOutput` necessary at all? I would expect that to be provided by `launch_testing`. Otherwise each package implementing test would need to implement that logic.

The length of each actual test file is also super long and verbose. I don't see an obvious way to reduce that though.

> There's currently no way to setup launch tests in pure Python packages

I think it would be very important to make it possible to write a test function which can run a launch test. The fact that this PR moves all the tests into a separate package is a good indicator that not having that functionality is resulting in new code to be written which immediately introduces tech debt.

**Comment by hidmic** [comment]
> I was wondering why are `TestCommandFinishesInAFiniteAmountOfTime` as well as TestCommandOutput necessary at all? I would expect that to be provided by launch_testing. Otherwise each package implementing test would need to implement that logic.

If we like this way of specifying tests we may standardize it, but no, it's not provided by `launch_testing` out of the box.

> I think it would be very important to make it possible to write a test function which can run a launch test. 

I agree with that, though it's not a blocker for these tests to land.

File: test_ros2cli/test/ros2msg_test_configuration.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            expected_output=lambda lines: 'std_msgs' in lines
+        ),
+        CLITestConfiguration(
+            command='msg',
+            arguments=['show', 'std_msgs/msg/String'],
+            expected_output=['string data'],
+        ),
+        CLITestConfiguration(
+            command='msg',
+            arguments=['show', 'std_msgs/msg/NotAMessageType'],
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply test_ros2cli/test/ros2msg_test_configuration.py]: and this one is fun:
```
ros2 msg show std_msgs/asd/String
string data
```

File: test_ros2cli/test/test_cli.py.in [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    ])
+    return launch_description, locals()
+
+
+class TestCommandFinishesInAFiniteAmountOfTime(unittest.TestCase):
+
+    def @TEST_NAME@(self, proc_info, command_under_test, test_configuration):
+        """Test that the command under test finished in a finite amount of time."""
+        success = proc_info.waitForShutdown(
+            process=command_under_test, timeout=test_configuration.timeout
```
Comment (ivanpauno) [inline test_ros2cli/test/test_cli.py.in]: `waitForShutdown` could take an optional keyword argument: `assert: bool = False`. In that way, we wouldn't need the extra `if ...: assert ...` here.
I think it may be useful.
Note: We have also `assertWaitForShutdown`, but it's less versatile than an extra argument.
Note: I won't delete the original `waitForShutdown`

File: test_ros2cli/test/ros2service_test_configuration.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+def get_test_configurations(rmw_implementation):
+    return [
+        CLITestConfiguration(
+            command='service',
+            arguments=['list'],
+            fixture_actions=[get_add_two_ints_server_action(),
+                             get_add_two_ints_server_action(service_name='_add_two_ints')],
+            expected_output=itertools.chain(
+                # Cope with launch internal ROS 2 node.
+                itertools.repeat(re.compile(r'/launch_ros/.*parameter.*'), 6),
```
Comment (ivanpauno) [inline test_ros2cli/test/ros2service_test_configuration.py]: We are checking for an exact number of lines here (6), of something that may easily change.
I don't know if it's better checking just if some lines are in the output, or checking for exact output.
This apply in many places, but I don't have an strong opinion about what's better.

Check lines in output pro: Updating the test is needed less often
Check lines in output con: Maybe, the test should have been updated (adding extra checks) after a change in other place, and it wasn't.

File: test_ros2cli/test/ros2topic_test_configuration.py [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+            output_filter=basic_output_filter(
+                filtered_rmw_implementation=rmw_implementation
+            ),
+            self_terminates=False,
+            exit_codes=[2],
+            timeout=5.0,
+        ),
+        CLITestConfiguration(
+            command='topic',
+            arguments=['pub', '/my_ns/chatter', 'std_msgs/msg/String', '{data: test}'],
```
Comment (ivanpauno) [inline test_ros2cli/test/ros2topic_test_configuration.py]: Should the test also subscribe to the topic and check that the message is received?
I think that checking output is not enough in this case

**Comment by ivanpauno** [comment]
> The size of the tests makes me wonder if we really can't write them more compact? Is there anything we could do API / feature wise which would help reduce this?

I think that most of the code is just test cases for parameterizing the test.
It will be difficult to do it more compact. 

> Instead we should find a way to write the test to work without this delay.

I agree with this point. It's probably a bit difficult to implement.

> I was wondering why are `TestCommandFinishesInAFiniteAmountOfTime` as well as `TestCommandOutput` necessary at all?

I did a comment about improving `waitForShutdown`, so `TestCommandFinishesInAFiniteAmountOfTime` will be shorter.
`launch_testing` methods for asserting process output could be better. Probably, something similar to `TestCommandOutput` could be provided by `launch_testing`.

File: test_ros2cli/test/ros2msg_test_configuration.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test_ros2cli/test/ros2msg_test_configuration.py]: That's interesting.

File: test_ros2cli/test/ros2service_test_configuration.py (hunk 9 above)
Comment (hidmic) [reply test_ros2cli/test/ros2service_test_configuration.py]: > Check lines in output con: Maybe, the test should have been updated (adding extra checks) after a change in other place, and it wasn't.

 That's the main reason why I went for a restrictive rather than a permissive test. If we trust CLI release testing on these tests, I'm more comfortable being as precise as possible regarding how they should behave.

File: test_ros2cli/test/ros2topic_test_configuration.py (hunk 10 above)
Comment (hidmic) [reply test_ros2cli/test/ros2topic_test_configuration.py]: That's what we do with the listener and we then check it's output. Yes, we're trusting the listener in the process...

**Comment by hidmic** [comment]
> It will be difficult to do it more compact.

That is true @dirk-thomas. Test parameterizations are as few and compact as they can be. We simply have way to many combinations that need to be tested. But the test itself could use an API simplification.

File: test_ros2cli/test/ros2msg_test_configuration.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test_ros2cli/test/ros2msg_test_configuration.py]: @ivanpauno I can confirm interface introspection CLIs ignore the interface type namespace. I've commented out those tests, we should address that in a follow up PR.

File: ros2node/ros2node/verb/list.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply ros2node/ros2node/verb/list.py]: I believe they show in discovery order, which is non-deterministic, yes.

**Comment by hidmic** [comment]
@ivanpauno @dirk-thomas rebased and ready for re-review.  https://github.com/ros2/ros2cli/pull/304/commits/bf24f0db7d2d578ceebf57fa01b51111cb28b76f is a *major* refactor of this PR, so anything before that commit is likely outdated. PTAL when you have time.

File: ros2topic/test/fixtures/controller_node.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+
+from geometry_msgs.msg import TwistStamped
```
Comment (dirk-thomas) [inline ros2topic/test/fixtures/controller_node.py]: Why using `geometry_msgs` for this test? Something from `std_msgs` would be a much lighter dependency.

Comment (hidmic) [reply ros2topic/test/fixtures/controller_node.py]: `ros2 topic delay` works with message types that have a header (with a stamp) and `std_msgs` doesn't have any. There are some in `visualization_msgs`, `diagnostic_msgs`, `stereo_msgs`, `sensor_msgs`, `trajectory_msgs` and `nav_msgs`. Would you rather switch to one of those? I don't have a strong opinion.

**Comment by hidmic** [comment]
CI (full, against Fast-RTPS, Connext and Opensplice)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8398/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4313/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6837/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8297/)

**Comment by hidmic** [comment]
> Beside the inline question about geometry_msgs how about the TODOs? Can you comment on why they are there?

Some of the tools do not behave as one would expect and I didn't want to block this PR on those fixes. Once those are fixed (or the tools are deprecated, https://github.com/ros2/ros2cli/pull/363), we can remove those TODOs.

File: ros2action/test/fixtures/fibonacci_action_server.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    except BaseException:
+        print('exception in server:', file=sys.stderr)
+        raise
+    finally:
+        node.destroy_node()
+        rclpy.shutdown()
+
+
+if __name__ == '__main__':
+    main()
```
Comment (ivanpauno) [inline ros2action/test/fixtures/fibonacci_action_server.py]: Maybe, we could provide a package with test nodes, so they can be used from other packages.
Probably, this is out of the scope of this PR.

File: ros2interface/ros2interface/api/__init__.py [hunk 13]
```diff
@@ -41,6 +41,8 @@ def get_interfaces(package_name):
 
 def get_interface_path(parts):
     prefix_path = has_resource('packages', parts[0])
+    if not prefix_path:
+        raise LookupError('Unknown package {}'.format(parts[0]))
```
Comment (ivanpauno) [inline ros2interface/ros2interface/api/__init__.py]: Bug fixes should be done in separated PRs.

File: ros2action/test/fixtures/fibonacci_action_server.py (hunk 12 above)
Comment (hidmic) [reply ros2action/test/fixtures/fibonacci_action_server.py]: I entertained that idea for a bit, but besides basic talker / listeners, fixture nodes tend to be quite specific to a test and not so easy to reuse unless you start generalizing them -- which I purposefully didn't want to.

File: ros2interface/ros2interface/api/__init__.py (hunk 13 above)
Comment (hidmic) [reply ros2interface/ros2interface/api/__init__.py]: Yes! I thought I had removed them all, looks like I didn't.

Comment (hidmic) [reply ros2interface/ros2interface/api/__init__.py]: See https://github.com/ros2/ros2cli/pull/366

**Comment by hidmic** [comment]
Considering how CI looks like atm, I don't think we'll get to merge this patch today. There are test fixture executables failing on specific platforms in a variety of ways, some even segfault in MacOS, while nightlies are all green. That means this is going to require further investigation on each platform.

CC @mjcarroll @dirk-thomas

**Comment by ivanpauno** [comment]
IMO, this can be merged after the feature freeze. It's just adding tests, no features.
It would be good to merge https://github.com/ros2/launch/pull/279 if it's ready, though.

**Comment by hidmic** [comment]
> It would be good to merge ros2/launch#279 if it's ready, though.

The same applies to that one. Test failures are mostly unrelated to this PR, but to other packages that make use of `launch_testing` and now fail in ways that are apparently unrelated to ros2/launch#279.

**Comment by hidmic** [comment]
Alright, recent builds from `master`  show that segmentation faults on MacOS w/ Opensplice (http://ci.ros2.org/job/ci_osx/6892/) and flaky pendulum control tests on Linux-aarch64 w/ Opensplice (http://ci.ros2.org/job/ci_linux-aarch64/4368/) are already there and unrelated to this patch. 

CI also experienced some issues and thus the current failed builds. I'll re-trigger after solving some other issues related to `ros2cli` tests on MacOS.

**Comment by hidmic** [comment]
Rebased to resolve conflicts.

**Comment by hidmic** [comment]
CI (full, against Fast-RTPS, Connext and Opensplice)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8490/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4378/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6905/) (skipping `ros2bag`)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8394/) (skipping `ros2bag`)

**Comment by hidmic** [comment]
Alright, @ivanpauno @mjcarroll may I ask for a final review? Test failures on Linux and MacOS are all unrelated to this patch and I had to re-trigger Windows CI because it failed for unrelated reasons (yet again). I'd like to merge this ASAP to avoid further conflicts with any new changes to `ros2/ros2cli` .

**Comment by ivanpauno** [comment]
Please, double check that all CI failures are unrelated.

**Comment by hidmic** [comment]
Alright, finally green! Though that means there's some flakiness that needs to be addressed. @mjcarroll may I proceed with this PR and any follow-up ones after it?
